File: apps/tabtin-android/app/src/main/java/com/tabtin/mobile/features/conversation/MessageQuote.kt
package com.tabtin.mobile.features.conversation

import com.tabtin.mobile.data.model.ChatMessage

internal data class ComposerMessageQuote(
    val author: String,
    val content: String,
    val reply: String,
    val payload: String,
)

internal object MessageQuote {
    fun payload(message: ChatMessage): String? {
        if (message.isStreaming) return null
        val content = message.displayContent.trim()
        if (content.isEmpty()) return null
        val author = if (message.isAssistant) "Agent" else "我"
        val quoted = content.lineSequence().joinToString("\n") { "> $it" }
        return "> $author：\n$quoted\n\n"
    }

    fun parseComposerDraft(draft: String): ComposerMessageQuote? {
        val separator = draft.indexOf("\n\n")
        if (separator < 0) return null

        val quoteBlock = draft.substring(0, separator)
        val lines = quoteBlock.split("\n")
        val author = when (lines.firstOrNull()) {
            "> Agent：" -> "Agent"
            "> 我：" -> "我"
            else -> return null
        }
        val contentLines = lines.drop(1).map { line ->
            when {
                line.startsWith("> ") -> line.removePrefix("> ")
                line == ">" -> ""
                else -> return null
            }
        }
        if (contentLines.isEmpty()) return null

        return ComposerMessageQuote(
            author = author,
            content = contentLines.joinToString("\n"),
            reply = draft.substring(separator + 2),
            payload = draft.substring(0, separator + 2),
        )
    }

    fun replacingComposerQuote(draft: String, message: ChatMessage): String? {
        val newPayload = payload(message) ?: return null
        return newPayload + removingComposerQuote(draft)
    }

    fun removingComposerQuote(draft: String): String = parseComposerDraft(draft)?.reply ?: draft
}

File: apps/tabtin-ios/Tabtin/Features/Conversation/MessageQuote.swift
import Foundation

struct ComposerMessageQuote: Equatable {
    let author: String
    let content: String
    let reply: String
    let payload: String
}

enum MessageQuote {
    static func payload(for message: ChatMessage) -> String? {
        guard !message.isStreaming else { return nil }
        let content = message.text.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !content.isEmpty else { return nil }
        let author = message.role == .assistant ? "Agent" : "我"
        let quoted = content.split(separator: "\n", omittingEmptySubsequences: false)
            .map { "> \($0)" }
            .joined(separator: "\n")
        return "> \(author)：\n\(quoted)\n\n"
    }

    static func parseComposerDraft(_ draft: String) -> ComposerMessageQuote? {
        guard let separator = draft.range(of: "\n\n") else { return nil }

        let quoteBlock = String(draft[..<separator.lowerBound])
        let lines = quoteBlock.split(separator: "\n", omittingEmptySubsequences: false).map(String.init)
        let author: String
        switch lines.first {
        case "> Agent：": author = "Agent"
        case "> 我：": author = "我"
        default: return nil
        }

        let contentLines = lines.dropFirst().compactMap { line -> String? in
            if line.hasPrefix("> ") { return String(line.dropFirst(2)) }
            if line == ">" { return "" }
            return nil
        }
        guard contentLines.count == lines.count - 1, !contentLines.isEmpty else { return nil }

        return ComposerMessageQuote(
            author: author,
            content: contentLines.joined(separator: "\n"),
            reply: String(draft[separator.upperBound...]),
            payload: String(draft[..<separator.upperBound])
        )
    }

    static func replacingComposerQuote(in draft: String, with message: ChatMessage) -> String? {
        guard let newPayload = payload(for: message) else { return nil }
        return newPayload + removingComposerQuote(from: draft)
    }

    static func removingComposerQuote(from draft: String) -> String {
        parseComposerDraft(draft)?.reply ?? draft
    }
}
